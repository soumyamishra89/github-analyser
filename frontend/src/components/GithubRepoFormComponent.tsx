import React from 'react';
import api from '../services/remote/api';
import { Form, Button, Col, Row } from 'react-bootstrap';
import { connect } from 'react-redux';
import { mapGithubDispatchToProps } from '../redux/dispatchMapping';
import { AppState } from '../types';

interface Props {
    loadGithubRepos: () => void,
    isGithubReposLoading: boolean;
}

interface State {
    owner: string,
    reponame: string,
    validated?: boolean,
    loading?: boolean
} 

export class GithubRepoFormComponent extends React.PureComponent<Props, State> {

    state: State = {
        owner: '',
        reponame: '',
        loading: false
    };

    componentDidUpdate(prevProps: Props) {
        if (prevProps.isGithubReposLoading === true && this.props.isGithubReposLoading === false) {
            this.setState({loading: false})
        }
    }

    postGithubInfo = async (event: any) => {
        event.preventDefault();
        const form = event.currentTarget;
        if (form.checkValidity() === false) {
            event.stopPropagation();
        } else {
            const { owner, reponame } = this.state;
            if (owner && reponame) {
                this.setState({loading: true});
                const resp = await api.postGithubInfo(owner, reponame);
                if (resp.status === 202) {
                    this.props.loadGithubRepos();
                } else {
                    this.setState({loading: false});
                    alert('Invalid github repo');
                }
            }
        }

        this.setState({validated: true});
    }

    onOwnerChange = (event: any) => {
        event.preventDefault();
        const value = event.target.value;
        this.setState({owner: value, validated: false});
    }

    onRepoChange = (event: any) => {
        event.preventDefault();
        const value = event.target.value;
        this.setState({reponame: value, validated: false});
    }

    render() {     
        
        return (
            <Form noValidate validated={this.state.validated} onSubmit={this.postGithubInfo}>
                <Form.Group>
                    <Form.Row>
                        <Col >
                            <Form.Control placeholder="Owner" required onChange={this.onOwnerChange} value={this.state.owner}/>
                        </Col>
                        <Col>
                            <Form.Control placeholder="Repo name" required onChange={this.onRepoChange} value={this.state.reponame}/>
                        </Col>
                        <Col>
                        <Button variant={this.state.loading ? "secondary" : "primary"} type="submit" className="position-relative d-flex justify-content-center align-items-center" disabled={this.state.loading}>
                            {this.state.loading && <div className="position-absolute button-loader"></div>}
                            Submit
                        </Button>
                        </Col>
                    </Form.Row>

                </Form.Group>
                <div className="d-flex justify-content-center">
                
                </div>
            </Form>
        )
    }
}

const mapStateToProps = (state: AppState) => ({
    isGithubReposLoading: state.isGithubReposLoading
});

export default connect(mapStateToProps, mapGithubDispatchToProps)(GithubRepoFormComponent);