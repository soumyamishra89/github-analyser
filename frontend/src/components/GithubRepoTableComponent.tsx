import React from 'react';
import { GithubRepo, AppState } from '../types';
import { Table, Modal } from 'react-bootstrap';
import { connect } from 'react-redux';
import Markdown from 'react-markdown';

interface Props {
    githubRepos: GithubRepo[]
}

interface State {
    readmeToShow?: string
}

class GithubRepoTableComponent extends React.PureComponent<Props, State> {

    state: State = {};

    showReadme = (evt: any, readme?: string) => {
        evt.preventDefault();
        if (!readme) {
            alert('No readme to show');
        } else {
            this.setState({readmeToShow: atob(readme)});
        }
    }

    hideReadme = () => {
        this.setState({readmeToShow: undefined});
    }

    render() {
        const { githubRepos } = this.props;
        if (githubRepos.length === 0) {
        return (<div className="d-flex justify-content-center align-item-center">
                Add a repository from github to show analysis
            </div>);
        }
        
        return (
            <>
                <table className="ml-2 mr-2 table table-striped table-bordered table-hover table-layout" >
                    <thead>
                        <tr>
                            <th>Owner</th>
                            <th>Repo name</th>
                            <th>Url</th>
                            <th>Commits</th>
                            <th>Open pull requests</th>
                            <th>Readme</th>
                        </tr>
                    </thead>
                    <tbody>
                        {githubRepos.map((gr, index) => (<tr key={index}>
                            <td>{gr.owner}</td>
                            <td>{gr.name}</td>
                            <td><a href={gr.url}>{gr.url}</a></td>
                            <td>{gr.commits}</td>
                            <td>{gr.openPullRequests}</td>
                            <td><button className="link-button" onClick={(evt: any) => this.showReadme(evt, gr.readme)}>Show Readme</button></td>
                        </tr>))}                    
                    </tbody>
                </table>
                {this.state.readmeToShow && <ReadmeModal onHide={this.hideReadme} readme={this.state.readmeToShow}/>}
            </>
        )
    }
}

interface ReadmeProps {
    onHide: () => void,
    readme: string
}

function ReadmeModal(props: ReadmeProps) {
    return (
        <Modal
            show={true}
            onHide={props.onHide}
            size="lg"
            aria-labelledby="contained-modal-title-vcenter"
            centered
        >
            <div className="pl-2 pr-2">
                <Markdown source={props.readme} />
            </div>
        </Modal>
    )
}

const mapStateToProps = (state: AppState) => ({
    githubRepos: state.githubRepos
});

export default connect(mapStateToProps)(GithubRepoTableComponent);