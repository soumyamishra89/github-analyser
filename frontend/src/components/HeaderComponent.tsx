import React from 'react';
import { Navbar, Nav } from 'react-bootstrap';
import { ThunkDispatch } from 'redux-thunk';
import { AppState } from '../types';
import PackageJson from '../../package.json';
import api from '../services/remote/api';
import { connect } from 'react-redux';
import GithubRepoFormComponent from './GithubRepoFormComponent';

interface Props {
    pingLoginStatus: () => void
}
class HeaderComponent extends React.PureComponent<Props, any> {

    componentDidMount() {
       this.props.pingLoginStatus();
    }

    render() {     
        
        return(
            <Navbar fixed="top" bg="dark" variant="dark" expand="lg">                   
                <Navbar.Brand>
                    Github Analyser <sup>{PackageJson.version}</sup>
                </Navbar.Brand>
                <Navbar.Toggle />
                <Navbar.Collapse>
                    <Nav className="mr-auto" />   
                    <GithubRepoFormComponent />
                </Navbar.Collapse>
            </Navbar>
        );
    }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, any, any>) => ({
    pingLoginStatus: () => dispatch(api.pingLoginStatus())
})

export default connect(null, mapDispatchToProps)(HeaderComponent);