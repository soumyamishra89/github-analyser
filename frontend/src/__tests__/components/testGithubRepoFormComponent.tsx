import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import {GithubRepoFormComponent} from '../../components/GithubRepoFormComponent';

const createProps = () => ({
    isGithubReposLoading: false,
    loadGithubRepos: jest.fn()
})

describe("Test GithubRepoFormComponent rendering", () => {
    let props;
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        props = createProps();
        
        wrapper = shallow(<GithubRepoFormComponent {...props}/>);
    });

    it("Snapshot test without loading", () => {
        expect(wrapper).toMatchSnapshot("Without Loading");
    });

    it("Snapshot test with loading", () => {
        wrapper.setState({loading: true});
        expect(wrapper).toMatchSnapshot("With Loading");
    });
})