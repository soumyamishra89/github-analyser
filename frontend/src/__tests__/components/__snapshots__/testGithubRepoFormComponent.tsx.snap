// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GithubRepoFormComponent rendering Snapshot test with loading: With Loading 1`] = `
<Form
  inline={false}
  noValidate={true}
  onSubmit={[Function]}
>
  <FormRow>
    <Col>
      <FormControl
        onChange={[Function]}
        placeholder="Owner"
        required={true}
        value=""
      />
    </Col>
    <Col>
      <FormControl
        onChange={[Function]}
        placeholder="Repo name"
        required={true}
        value=""
      />
    </Col>
    <Col>
      <Button
        active={false}
        className="position-relative d-flex justify-content-center align-items-center"
        disabled={true}
        type="submit"
        variant="secondary"
      >
        <div
          className="position-absolute button-loader"
        />
        Submit
      </Button>
    </Col>
  </FormRow>
</Form>
`;

exports[`Test GithubRepoFormComponent rendering Snapshot test without loading: Without Loading 1`] = `
<Form
  inline={false}
  noValidate={true}
  onSubmit={[Function]}
>
  <FormRow>
    <Col>
      <FormControl
        onChange={[Function]}
        placeholder="Owner"
        required={true}
        value=""
      />
    </Col>
    <Col>
      <FormControl
        onChange={[Function]}
        placeholder="Repo name"
        required={true}
        value=""
      />
    </Col>
    <Col>
      <Button
        active={false}
        className="position-relative d-flex justify-content-center align-items-center"
        disabled={false}
        type="submit"
        variant="primary"
      >
        Submit
      </Button>
    </Col>
  </FormRow>
</Form>
`;
